### PR TITLE
Add TTL on S3 CAS uploads

### PR DESCRIPTION
**Pull request by DaveCTurner**
Compare-and-swap operations on a S3 repository are implemented using
multipart uploads. Today to try and avoid collisions we refuse to
perform a compare-and-swap if there are other concurrent uploads in
progress. However this means that a node which crashes partway through a
compare-and-swap will block all future register operations.

With this commit we introduce a time-to-live on S3 multipart uploads,
such that uploads older than the TTL now do not block future
compare-and-swap attempts.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Closes ES-6567

**Review by henningandersen — APPROVED** [review]
LGTM.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java [hunk 1]
```diff
@@ -674,12 +693,41 @@ private int getUploadIndex(String targetUploadId, List<MultipartUpload> multipar
...
+         */
+        private boolean checkPreexistingUploads() {
+            final var uploads = listMultipartUploads();
+            if (uploads.isEmpty()) {
+                return false;
+            }
+
+            final var expiryDate = Date.from(
+                Instant.ofEpochMilli(
+                    blobStore.getThreadPool().absoluteTimeInMillis() - blobStore.getCompareAndExchangeTimeToLive().millis()
```
Comment (fcofdez) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:707]: I guess we can tolerate a clock drift between the S3 server and the local node clock in this case? the default expiry time is relatively high.

Comment (fcofdez) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:707]: Aha, I've just seen the log message that warns about this potential issue

**Review by fcofdez — APPROVED** [review]
LGTM

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java [hunk 2]
```diff
@@ -784,6 +824,16 @@ void run(BytesReference expected, BytesReference updated, ActionListener<Optiona
             }
         }
 
+        private void safeAbortMultipartUpload(String uploadId) {
+            try {
+                abortMultipartUploadIfExists(uploadId);
+            } catch (Exception e) {
+                // cleanup is a best-effort thing, we can't do anything better than log and fall through here
+                logger.error("unexpected error cleaning up upload [" + uploadId + "] of [" + blobKey + "]", e);
+                assert false : e;
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java]: Yes you're right we can't assert here - removed in 58d6b16db8000ea91d8286cd84a0e5e4edd2a9ad.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Service.java [hunk 3]
```diff
@@ -51,6 +54,12 @@
 class S3Service implements Closeable {
     private static final Logger LOGGER = LogManager.getLogger(S3Service.class);
 
+    private static final Setting<TimeValue> REPOSITORY_S3_CAS_TTL_SETTING = Setting.timeSetting(
+        "snapshot.repository_s3.compare_and_exchange.time_to_live",
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Service.java]: Sure, 869afe3f859a371ff47f0e3ef8146a4ad8a5ee6d

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java [hunk 4]
```diff
@@ -674,12 +693,41 @@ private int getUploadIndex(String targetUploadId, List<MultipartUpload> multipar
             return found ? uploadIndex : -1;
         }
 
-        void run(BytesReference expected, BytesReference updated, ActionListener<OptionalBytesReference> listener) throws Exception {
+        /**
+         * @return {@code true} if there are already ongoing uploads, so we should not proceed with the operation
+         */
+        private boolean checkPreexistingUploads() {
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java]: Sure, renamed in 431cbb2bfd6cfa930059ad12c9b818502c70ebcf.

**Pull request merged**